### DISCUSSION TITLE
Regarding GameboardController undo move

### DISCUSSION THREAD

**Discussion by DelusionArun** [CastleLab/COMP3021-2021Fall-PA1-Student-Version#13] [Q&A] [answered]
After undo, I know that the reverted player must be the same player. For example, if the original player was player@212 in some board @.&ast#, and I moved to ...@, then reverting will lead to @..&ast# where player is still player@212.

But I am a bit confused about the gems? If the original gem was Gem@12, should the reverted gem also be Gem@12 or is any gem fine? 

I think any gem/entity should be fine since there is no getGem/getEntity in gameboard for testing, but I want to make sure.

**Comment by Derppening** [comment, marked as answer]
Yes, any gem or entity is fine when you are restoring gems and entities (except for the player, as you said). It need not to be the same instance of gem or entity.

**Reply by DelusionArun** [reply]
Thank you for clarifying

**Comment by DelusionArun** [comment]
I also have a bit of confusion about the undo sequences. Although my undo test cases for both controllers pass, I wonder if my understanding below correct?

- The undoMove in GameBoardController only undoes the movement in the board (for example revert . to * or smt similar), but not the states.
- The processUndo in the GameController reverts the GameState (gamestate elements) and also reverts board by using GameBoardController. (I am mostly confused about this part. Because even without modifying gameboard, test cases seem to pass. But I saw that getNumGem was called in testing of processMove and getNumGems in gameboard is mutated by GameBoardController.)

**Reply by Derppening** [reply]
- `GameBoardController.undoMove` undoes all changes you have made *to the game board* for the specified move
- `GameController.processUndo` undoes all changes for a specified move, including anything you have changed in the `GameState`

Remember that there are hidden test cases, so test cases may pass even if your implementation is not fully correct.
